guides/2021-05-02-build-flags-backwards-compatibility/en.markdown

```diff
@@ -0,0 +1,163 @@
+
+Sometimes, when a new `go` version is released, it also ships with a bunch of changes and really interesting features on the standard library.
+As the time of this article, [go 1.16](https://golang.org/doc/go1.16) has been released around two months ago which introduces some changes and
+new features into the [core library](https://golang.org/doc/go1.16#library) like the new `io.FS` interface, the `go:embed` directive amongst others.
+
+As a module author, how could I introduce these new features and at the same time provide some guarantees that
+my module can still support the last N releases of go?
+
+This guide explains how to deal with situations where you want to use new features of recent versions of `go`
+and at the same time you don't want to force downstream dependecies to upgrade. In this case we'll be using
+conditional compilation through build tags in a real case scenario by using `go` 1.15 and 1.16 new `io.FS`
+package respectively.
+
+Here's a high level overview of the steps you'll accomplish following this guide:
+
+- Create a `{{{ .public }}}` module that contains and exports a `{{{ .public_func }}}` function
+- Create a `{{{ .gopher }}}` module that uses the `{{{ .public }}}`module
+- Bump the `{{{ .public }}}` module to use `go` 1.16 features without any considerations
+- Try updating the `{{{ .gopher }}}` module and show the observed errors
+- Fix the `{{{ .public }}}` module by adding required buid tags.
+
+
+### A simple go 1.15 program using ioutil.Discard
+
+Verifying that we're using `go` 1.15 version:
+
+{{{ step "go115version" }}}
+
+Now, we'll also check that we have `go` 1.16 installed as well:
+
+{{{ step "go116version" }}}
+
+
+We'll start by setting `go` 1.15 as the default working version:
+
+{{{ step "go115default" }}}
+
+Start by a new `public` module:
+
+{{{ step "public_init" }}}
+
+Now, we'll upload a simple `go` program that uses 1.15 `ioutil.Discard` in a public function:
+
+Create an initial version of the `{{{ .public_func }}}` in `{{{ .public_go }}}`:
+
+{{{ step "public_go_initial" }}}
+
+Commit and push this initial version:
+
+{{{ step "public_initial_commit" }}}
+
+### The `{{{ .gopher }}}` module
+
+Now create a `{{{ .gopher }}}` module to try out the `{{{ .public }}}` module. Unlike
+the `{{{ .public }}}` module, you will not publish the `{{{ .gopher }}}` module; it
+will be local only:
+
+{{{ step "gopher_init" }}}
+
+Create an initial version of a `main` package that uses the previous public module, in `{{{ .gopher_go }}}`:
+
+{{{ step "gopher_go_initial" }}}
+
+
+Get the initial version of the `{{{.gopher}}}` module:
+
+{{{ step "gopher_get" }}}
+
+Now, we'll run the `{{{ .gopher }}}` module `main` package:
+
+{{{ step "gopher_run" }}}
+
+
+### Bumping the `{{{ .public }}}` dependency to `go` 1.16
+
+`Go` 1.16 has been released, and as good developers we want to refactor our
+code so it uses the new `io.Discard` variable instead of the deprecated `ioutil` one.
```

Comment:
The thing that is really good about the example that you've chosen is that decision to use `io.Discard` could, on the face of things, be considered an entirely internal refactoring decision because it's not part of the API surface of your module.

Reply:
:hugs: